Fix riak_kv_pb_object hack and kill dialyzer warnings [JIRA: RIAK-3240]

nickelization opened the pull request:
Someone had previously suppressed some dialyzer warnings in riak_kv_pb_object, which were arising due to an ugly hack where one of the protobuf put request fields was being abused to pass info to successive recursive calls to `process`.

This commit fixes riak_kv_pb_object to use a state flag for this purpose instead of screwing with the request record, and also removes the corresponding lines from dialyzer.ignore-warnings since the warnings no longer exist.

nickelization committed:
Fix riak_kv_pb_object hack and kill dialyzer warns

Someone had previously suppressed some dialyzer warnings in
riak_kv_pb_object, which were arising due to an ugly hack where one of
the protobuf put request fields was being abused to pass info to
successive recursive calls to `process`.

This commit fixes riak_kv_pb_object to use a state flag for this purpose
instead of screwing with the request record, and also removes the
corresponding lines from dialyzer.ignore-warnings since the warnings no
longer exist.

lukebakken commented on src/riak_kv_pb_object.erl:
```diff
@@ -190,8 +193,8 @@ process(#rpbputreq{bucket=B0, type=T, key=K, vclock=PbVC,
             process(Req#rpbputreq{if_not_modified=undefined,
-                                  if_none_match=consistent},
-                    State);
+                                  if_none_match=undefined},
```
It seems weird to set `if_none_match` to `undefined` here, but then later set it to `true` for the `put` options.

nickelization replied:
Yeah, I agree it's not pretty. Still an improvement from the way it was though IMO, and I wanted to fix one hack without getting sucked into fixing all the other ugliness in this function. Trouble is, we need it to hit the next `process` clause after recursing, and if we leave `if_none_match` as true, it's going to loop back into the same clause again, so I think this is the best we can do without embarking on a big refactoring project.

nickelization replied:
Thought about this some more. We could add a `State#state.consistent == false` clause in that function head to force it to fall down to the next clause after we set that flag. Not sure if this would be that much better, since we're adding even more complexity to the control flow, and we never look at `if_not_modified` again anyway...but if that would address your concerns I could see that as a valid option.

lukebakken replied:
Even though it's ignored, why not set `if_none_match` to `true` here, with a comment as to why you're setting it to true:

```erlang
%% HAY Even though this is ignored it's getting set to true here so you
%% know that later on in process() it's going to get set to true in the
%% options passed to put()
```

I'm just thinking about how to reduce confusion for the next person (if any) who looks at `if_none_match=undefined` and thinks WTF

lukebakken replied:
Or, leave it as-is with a comment about later usage of `if_none_match`

nickelization replied:
Yeah good call, I'll add some comments. I didn't think it was that weird at first because there are other case clauses in the same function where we set `if_none_match` and `if_not_modified` to `undefined` before recursing, but I can see how it's more confusing in this particular case.

lukebakken replied:
Well, I didn't expand the code, which could have made it more obvious to me 😸

nickelization replied:
Haha, well it's done now anyway 😄 . Any other review comments, or are you ready to give it a +1?

nickelization committed:
Add comments about weird if_not_modified handling

What we're doing here is a bit counter-intuitive, so just adding
comments to help out the next person that comes along and looks at this
code. I would like to fix this more thoroughly, but don't have time to
refactor all of this code right now.

bashopatricia commented:
create jira issue

fadushin commented:
Probably worth pointing out that there is no behavioral change with this code change, so that there should therefore be no additional test required.

fadushin commented:
+1

nickelization committed:
Merge branch 'develop' into nem-fix-riak_kv_pb_object-consistent-hack
